lib/events/athena/publisher.go

```diff
@@ -0,0 +1,172 @@
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/sns"
+	snsTypes "github.com/aws/aws-sdk-go-v2/service/sns/types"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+)
+
+const (
+	payloadTypeAttr          = "payload_type"
+	payloadTypeRawProtoEvent = "raw_proto_event"
+	payloadTypeS3Based       = "s3_based_payload"
+
+	// maxSNSMessageSize defines maximum size of SNS message. AWS allows 256KB
+	// however it counts also headers. We round it to 250KB, just to be sure.
+	maxSNSMessageSize = 250 * 1024
+	// maxS3BasedSize defines some resonable threshold for S3 based messages (2GB).
+	maxS3BasedSize = 2 * 1024 * 1024 * 1024
+)
+
+// publisher is a SNS basesd events publisher.
+// It publishes proto events directly to SNS topic, or use S3 bucket
+// if payload is too large for SNS.
+type publisher struct {
+	topicARN      string
+	snsCli        snsCli
+	uploader      s3uploader
+	payloadBucket string
+	payloadPrefix string
+}
+
+type snsCli interface {
+	Publish(ctx context.Context, params *sns.PublishInput, optFns ...func(*sns.Options)) (*sns.PublishOutput, error)
+}
+
+type s3uploader interface {
+	Upload(ctx context.Context, input *s3.PutObjectInput, opts ...func(*manager.Uploader)) (*manager.UploadOutput, error)
+}
+
+// newPublisher returns new instance of publisher.
+func newPublisher(cfg Config, awsCfg aws.Config, log *log.Entry) (*publisher, error) {
+	url, err := url.Parse(cfg.LargeEventsS3)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// TODO(tobiaszheller): consider reworking lib/observability to work also on s3 sdk-v2.
+	return &publisher{
+		topicARN:      cfg.TopicARN,
+		snsCli:        sns.NewFromConfig(awsCfg),
+		uploader:      manager.NewUploader(s3.NewFromConfig(awsCfg)),
+		payloadBucket: url.Host,
+		payloadPrefix: strings.TrimSuffix(strings.TrimPrefix(url.Path, "/"), "/"),
+	}, nil
+}
+
+// EmitAuditEvent emits audit event.
+func (p *publisher) EmitAuditEvent(ctx context.Context, in apievents.AuditEvent) error {
+	// Just double check that audit event has minimum necessary fields for athena
+	// to works. Teleport emitter layer above makes sure that they are filled.
+	if in.GetID() == "" {
+		return trace.BadParameter("missing uid of audit event %s", in.GetType())
+	}
+	if in.GetTime().IsZero() {
+		return trace.BadParameter("missing time of audit event %s", in.GetType())
+	}
+
+	oneOf, err := apievents.ToOneOf(in)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	marshaledProto, err := oneOf.Marshal()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// TODO(tobiaszheller): verify if those limits and retry is valid.
+	retry, err := retryutils.NewLinear(retryutils.LinearConfig{
+		Step: 100 * time.Millisecond,
+		Max:  1 * time.Minute,
+	})
```

Comment:
@rosstimothy what do you think about those values? i think we can accept waiting for 1minute, but maybe I am wrong.

Reply:
good idea, I can use the one from AWS, what do you think about values [there](https://github.com/aws/aws-sdk-go-v2/blob/e2147ef38518dea0e345ee9bb0e244582ef9cdec/aws/retry/standard.go#L26)? 
> 	DefaultMaxAttempts int = 3
>	DefaultMaxBackoff time.Duration = 20 * time.Second

Is 1 minute good idea? default for aws is 20s.

Reply:
I don't think it will hurt us to extend those values a little bit. I think that dynamo offers SLA 99.99% while SNS 99.9% so it make sense to retry a little bit more.